beacon-chain/core/altair/epoch_precompute.go

```diff
@@ -0,0 +1,229 @@
+		// Was validator withdrawable or slashed
+		withdrawable := prevEpoch+1 >= val.WithdrawableEpoch()
+		pVal := &precompute.Validator{
+			IsSlashed:                    val.Slashed(),
+			IsWithdrawableCurrentEpoch:   withdrawable,
+			CurrentEpochEffectiveBalance: val.EffectiveBalance(),
+			InactivityScore:              inactivityScores[idx],
+		}
+		// Validator active current epoch
+		if helpers.IsActiveValidatorUsingTrie(val, helpers.CurrentEpoch(state)) {
+			pVal.IsActiveCurrentEpoch = true
+			bal.ActiveCurrentEpoch += val.EffectiveBalance()
+		}
+		// Validator active previous epoch
+		if helpers.IsActiveValidatorUsingTrie(val, prevEpoch) {
+			pVal.IsActivePrevEpoch = true
+			bal.ActivePrevEpoch += val.EffectiveBalance()
+		}
+
+		pValidators[idx] = pVal
+		return nil
+	}); err != nil {
+		return nil, nil, errors.Wrap(err, "could not initialize epoch validator")
+	}
+	return pValidators, bal, nil
+}
+
+// ProcessEpochParticipation processes the epoch participation in state and update individual validator's pre computes,
+// it also tracks and updates epoch attesting balances.
+func ProcessEpochParticipation(
+	ctx context.Context,
+	state iface.BeaconState,
+	bal *precompute.Balance,
+	vals []*precompute.Validator,
+) ([]*precompute.Validator, *precompute.Balance, error) {
+	ctx, span := trace.StartSpan(ctx, "altair.ProcessEpochParticipation")
+	defer span.End()
+
+	cp, err := state.CurrentEpochParticipation()
+	if err != nil {
+		return nil, nil, err
+	}
+	for i, b := range cp {
+		if HasValidatorFlag(b, params.BeaconConfig().TimelyTargetFlagIndex) {
+			vals[i].IsCurrentEpochTargetAttester = true
+		}
+	}
+	pp, err := state.PreviousEpochParticipation()
+	if err != nil {
+		return nil, nil, err
+	}
+	for i, b := range pp {
+		if HasValidatorFlag(b, params.BeaconConfig().TimelySourceFlagIndex) {
+			vals[i].IsPrevEpochAttester = true
+		}
+		if HasValidatorFlag(b, params.BeaconConfig().TimelyTargetFlagIndex) {
+			vals[i].IsPrevEpochTargetAttester = true
+		}
+		if HasValidatorFlag(b, params.BeaconConfig().TimelyHeadFlagIndex) {
+			vals[i].IsPrevEpochHeadAttester = true
+		}
+	}
+	bal = precompute.UpdateBalance(vals, bal)
+	return vals, bal, nil
+}
+
+// ProcessRewardsAndPenaltiesPrecompute processes the rewards and penalties of individual validator.
+// This is an optimized version by passing in precomputed validator attesting records and and total epoch balances.
+func ProcessRewardsAndPenaltiesPrecompute(
+	state iface.BeaconStateAltair,
+	bal *precompute.Balance,
+	vals []*precompute.Validator,
+) (iface.BeaconStateAltair, error) {
+	// Don't process rewards and penalties in genesis epoch.
+	if helpers.CurrentEpoch(state) == 0 {
+		return state, nil
+	}
+
+	numOfVals := state.NumValidators()
+	// Guard against an out-of-bounds using validator balance precompute.
+	if len(vals) != numOfVals || len(vals) != state.BalancesLength() {
+		return state, errors.New("registries not the same length as state registries")
```

Comment:
```suggestion
		return state, errors.New("validator registries not the same length as state's validator registries")
```
Just to make it clear which registries we are talking about.